synapse/events/snapshot.py

```diff
@@ -254,6 +299,135 @@ async def get_prev_state_ids(
+    state_delta_due_to_event: Optional[dict]
+    prev_group: Optional[int]
+    delta_ids: Optional[StateMap[str]]
+    partial_state: Optional[bool]
+    state_map: Optional[StateMap[str]] = None
+
+    async def get_prev_state_ids(
+        self, state_filter: Optional["StateFilter"] = None
+    ) -> StateMap[str]:
+        """
+        Gets the room state map, excluding this event.
+
+        Args:
+            state_filter: specifies the type of state event to fetch from DB
+
+        Returns:
+            Maps a (type, state_key) to the event ID of the state event matching
+            this tuple.
+        """
+        if self.state_map:
+            return self.state_map
+
+        assert self.state_group_before_event is not None
+        return await self._storage.state.get_state_ids_for_group(
+            self.state_group_before_event, state_filter
+        )
+
+    async def persist(self, event: EventBase) -> EventContext:
+        """
+        Creates a full `EventContext` for the event, persisting any referenced state that
+        has not yet been persisted.
+
+        Args:
+             event: event that the EventContext is associated with.
+
+        Returns: An EventContext suitable for sending to the database with the event
+        for persisting
+        """
+        assert self.partial_state is not None
+        # If we have a full set of state for at/before the event but don't have a state
+        # group for that state, we need to get one
+        if self.state_map:
+            if self.state_group_before_event is None:
+                state_group_before_event = await self._storage.state.store_state_group(
+                    event.event_id,
+                    event.room_id,
+                    prev_group=None,
+                    delta_ids=None,
+                    current_state_ids=self.state_map,
+                )
+                self.state_group_before_event = state_group_before_event
+                self.prev_group = state_group_before_event
+        if not event.is_state():
+            return EventContext.with_state(
+                storage=self._storage,
+                state_group=self.state_group_before_event,
+                state_group_before_event=self.state_group_before_event,
+                state_delta_due_to_event=self.state_delta_due_to_event,
+                partial_state=self.partial_state,
+                prev_group=self.prev_group,
+                delta_ids=self.delta_ids,
+            )
+        else:
+            if not self.prev_group and not self.delta_ids:
+                # if we don't have a prev_group and the delta from the prev_group we need
+                # to store the state group with the full state
+                state_group_after_event = await self._storage.state.store_state_group(
+                    event.event_id,
+                    event.room_id,
+                    prev_group=self.prev_group,
+                    delta_ids=self.delta_ids,
```

Comment:
Isn't `delta_ids` the delta to `state_map`, which is for the state *before* the event? I think this should be 

```suggestion
                    prev_group=self.state_group_before_event,
                    delta_ids=self.state_deltas_due_to_event,
```